---
title: "Metagenomics"
author: "Allison Harvey"
date: "6/29/2021"
output: html_document
---

```{r}
#install needed packages
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("phyloseq")
```

```{r}
library(tidyverse)
library(phyloseq)
```
Activate qiime2
```{bash}
conda activate qiime2-2021.4
source tab-qiime
```


Determine number of sequences present for each barcode
```{bash}

grep ">" ~/MetagenomicsLab/input/Data/RiceSeqs.fna | cut -d " " -f 4 | sort | uniq -c
```

__Exercise 1:__ Using information in the RiceMappingFile.txt and RiceSeqs.fna answer the following questions. Are the number of sequences for each sample approximately the same or are there any outliers? If so, which samples do they belong to? Could a different number of sequences per sample affect future analysis? Explain your reasoning.

Most are fairly similar but there are some samples that have a much higher or lower number of sequences than the others. Different numbers of sequences per sample could bias results towards certain samples in future analysis since outliers may appear more or less diverse due solely to the number of sequences present.

Low: 289 new_bc=AGCAGTCGCGATGT
High: 4880 new_bc=ACGGTGAGTGTCGT, 3870 new_bc=ACCGCAGAGTCAGT, 3092 new_bc=ACAGACCACTCAGT, 2965 new_bc=AACGCACGCTAGGT



Import data into qiime2 (run in terminal in input directory)
```{bash eval=FALSE}
qiime tools import \
  --input-path Data/RiceSeqs.fna \
  --output-path Data/RiceSeqs.qza \
  --type 'SampleData[Sequences]'
```

Dereplicate the sequences (run in terminal in input directory)
```{bash eval=FALSE}
qiime vsearch dereplicate-sequences \
  --i-sequences Data/RiceSeqs.qza \
  --o-dereplicated-table Data/RiceTable.qza \
  --o-dereplicated-sequences Data/RiceRep-seqs.qza
```

Cluster Microbiome sequences into OTUs (run in terminal in input directory)
  Cluster similar sequences
```{bash eval=FALSE}
qiime vsearch cluster-features-de-novo \
  --i-table Data/RiceTable.qza \
  --i-sequences Data/RiceRep-seqs.qza \
  --p-perc-identity 0.99 \
  --o-clustered-table Data/RiceTable-dn-99.qza \
  --o-clustered-sequences Data/RiceRep-seqs-dn-99.qza
```
  Match OTUs to reference database to assign taxonomy (5hrs) (run in terminal in input directory)
```{bash eval=FALSE}
time qiime feature-classifier classify-consensus-vsearch \
  --i-query Data/RiceRep-seqs-dn-99.qza \
  --i-reference-reads silva-138-99-seqs.qza \
  --i-reference-taxonomy silva-138-99-tax.qza \
  --p-threads 2 \
  --o-classification Data/RiceTaxTable.qza
```
  Export data in form that can be read by R phyloseq (run in terminal in input directory)
```{bash eval=FALSE}
#table of otus
qiime tools export \
  --input-path Data/RiceTable-dn-97.qza \
  --output-path qiime_export
  
#convert otus to text format
biom convert -i qiime_export/feature-table.biom -o qiime_export/otu_table.txt --to-tsv

# table of taxonomy
qiime tools export \
  --input-path Data/RiceTaxTable.qza \
  --output-path qiime_export

# sequences
qiime tools export \
  --input-path Data/RiceRep-seqs-dn-97.qza \
  --output-path qiime_export
  

```
  

__Exercise 2:__ From the OTU summary, look at how many OTUs correspond to each sample (“counts/sample detail”). Do technical replicates agree with one another? At this stage, what conclusions can you draw about the number of OTUs in these samples?

In some cases the technical replicates agree with one another but in other samples there is a larger difference between replicates. I'm not sure if we can make any conclusions based on this.


Import OTU table into R
```{r}
otu <- read.delim("../input/qiime_export/otu_table.txt", skip=1, row.names = 1, as.is = TRUE) %>% as.matrix()
head(otu)
```
```{r}
tax <- read.delim("../input/qiime_export/taxonomy.tsv", as.is = TRUE) %>%
  select(-Consensus) %>%
  mutate(Taxon = str_remove_all(Taxon, ".__| ")) %>%
  separate(Taxon, into = c("domain", "phylum", "class", "order", "family", "genus", "species"), sep = ";", fill="right")
rownames(tax) <- tax$Feature.ID
tax <- tax %>% select(-Feature.ID) %>% as.matrix()
tax[tax==""] <- NA
head(tax)
```

```{r}
#Create data frame of sample info
sampleinfo <- data.frame(sample=colnames(otu)) %>%
  mutate(cultivar=str_sub(sample,1,1),
         cultivar={str_replace(cultivar, "M", "M104") %>%
             str_replace( "I", "IR50") %>%
             str_replace( "N", "Nipponbarre")},
         time=str_extract(sample,"E|L"),
         location={str_extract(sample,".[12]") %>% str_sub(1,1)},
         location={str_replace(location, "B", "rhizoplane") %>%
             str_replace("M", "rhizosphere") %>%
             str_replace("E", "endosphere")})
rownames(sampleinfo) <- sampleinfo$sample
sampleinfo <- sampleinfo %>% select(-sample)
head(sampleinfo)
```

```{r}
#merge OTU and taxonomy info into new phyloseq object
rice.ps <- phyloseq(otu_table(otu,taxa_are_rows=TRUE), tax_table(tax), sample_data(sampleinfo))
rice.ps
```

```{r}
#Filter to remove rare sequences
rice.ps.small <- filter_taxa(rice.ps, function(x) sum(x > 1) > 2, TRUE) #require greater than one observation in more than two samples
rice.ps.small
```

```{r}
plot_heatmap(rice.ps.small)
```

__Exercise 3:__ Although, the resolution of the y-axis makes it difficult to read each OTU, it is still a valuable preliminary visualization. What types of information can you gain from this heat map? Are there any trends present at this stage with respect to the various samples?

The samples appear to fall into two broad groups with one group represented by the samples in  the middle having abundant reads in the OTUs at the center of the axis, and generally containing reads from  a greater number of OTUs that the other samples which primarily contian OTUs from the top and bottom of the axis.


```{r}
#make bar plot. color by phylum.
plot_bar(rice.ps.small, fill="phylum")
```
```{r}
#repeat removing black lines
pl <- plot_bar(rice.ps.small, fill="phylum")
pl + geom_col(aes(fill=phylum, color=phylum))
```


__Exercise 4:__

a. Make a bar plot with samples grouped by location. When comparing by location, which groups are the predominant phyla in the different samples? Are there any predominant groups unique to particular sample treatments?

Proteobacteria appears to be the predominant phyla in all samples. Actinobacteria seems to be present at a higher abundance in samples in the endosphere and rhizosphere. 

```{r}
pl2 <- plot_bar(rice.ps.small, fill="phylum", facet_grid = ~location)
pl2 + geom_col(aes(fill=phylum, color=phylum))
```


b. Make a bar plot with samples grouped by cultivar. When comparing by cultivar, are the predominant phyla consistent with those observed in Part A? Are there any predominant phyla unique to a specific cultivar? What does this indicate to you about the effect of the genotype and/or the effect of the treatment?

The same phylums are consistent with those in part A. There don't appear to be any predominant phyla unique to a specific cultivar suggesting that treatment has a greater effect than genotype on microbial communitites.

```{r}
pl3 <- plot_bar(rice.ps.small, fill="phylum", facet_grid = ~cultivar)
pl3 + geom_col(aes(fill=phylum, color=phylum))
```


Determine the diversity within a sample
```{r}
plot_richness(rice.ps, measures=c("Observed", "Chao1", "Shannon"))
```

__Exercise 5:__
Is there an alpha diversity metric that estimates the sample diversity differently than the other metrics? If so, which one is it?

Shannon diversity estimates the sample diversity differently than the others. A greater number of samples are reported as having higher diversity.

__Exercise 6:__
Look at the help file for `plot_richness` and plot the samples grouped either by cultivar or location. Do either of these categories seem to affect species diversity? Thinking back to the difference in read counts per sample, could the differences in diversity just be due to differences in sequencing depth?

Location affects species diversity with greater diversity in rhizpsphere samples. In some genotypes this could ge due to greater sequenceing depth. Diversity is not a strongly influenced by cultivar.

```{r}
plot_richness(rice.ps, x = "location", measures=c("Observed", "Chao1", "Shannon"))
plot_richness(rice.ps, x = "cultivar", measures=c("Observed", "Chao1", "Shannon"))
```


Visualize diversity between samples
```{r}
#computed distances and MDS coordinates
rice.ord.small <- ordinate(rice.ps.small, method="NMDS", distance="bray")
```

```{r}
pl4 = plot_ordination(rice.ps.small, rice.ord.small, type="samples", color="cultivar", shape="location") 
pl4 + geom_point(size=4)
```

__Exercise 7:__

a. Does cultivar or location appear to have more of an influence on the clustering?

location appears to more strongly influence clustering.

b. Which two locations are more similar to one another? Does that make biological sense?

the rhizpsplane and endosphere appear most similar which makes sense as these are physically the closest to each other.

__Exercise 8:__

Four Nipponbarre samples form a distinct group separated from all other samples. Replot the data (changing the plot aesthetics) to provide an explanation.

Early timepoint samples of Nipponbarre in the rhizoplane and endosphere separate from other samples with like locations, but late timepoint samples group with the expected location groups.

```{r}
pl4 = plot_ordination(rice.ps.small, rice.ord.small, type="samples", color="cultivar", shape="location", label = "time") 
pl4 + geom_point(size=2)
```

